### PR TITLE
Additional GUI Features

### LINKED ISSUE #172: Implement GUI to allow user to specify data storage location
The ability to specify a specific file path as the data storage location. With this feature, Jim can choose to store the data file in a local folder controlled by a cloud syncing service (e.g. dropbox), allowing him to access task data from multiple computers. 
@cs2103jan2016-v15-1j/developers we totally forgot about this >.<

### PR DESCRIPTION
**Pull request by SorataYuu**
#213 Must be Closed First

Closes #172 #194 #198 #216 #219 #220

### CODE REVIEW

File: src/main/java/cs2103/v15_1j/jimjim/controller/JJController.java [hunk 1]
```diff
@@ -37,6 +37,16 @@ public DataLists getMasterList() {
...
 	}
+	
+	@Override
+	public String getFilePath() {
+		return this.states.config.savePath;
+	}
+	
+	@Override
+	public void setFilePath(String filePath) {
+		this.states.storage.setSaveFile(filePath);
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/controller/JJController.java]: Nope shouldn't do this

Comment (SorataYuu) [inline src/main/java/cs2103/v15_1j/jimjim/controller/JJController.java]: How would you suggest changing the filePath then?

Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/controller/JJController.java]: Need to save to config and save it also

File: src/main/java/cs2103/v15_1j/jimjim/ui/BottomPaneController.java [hunk 2]
```diff
@@ -48,8 +53,10 @@
 	private final double EXECUTE_BTN_HEIGHT = 30.0;
 	private final double HELP_BTN_WIDTH = 30.0;
 
-	public BottomPaneController(MainViewController con){
+	public BottomPaneController(MainViewController con, Stage primaryStage, String filePath){
 		this.con = con;
+		this.primaryStage = primaryStage;
+		this.filePath = filePath;
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/BottomPaneController.java]: better to just keep asking the controller than to keep a version of your own and then not synchronized

File: src/main/java/cs2103/v15_1j/jimjim/ui/BottomPaneController.java [hunk 3]
```diff
@@ -211,6 +215,26 @@ else if (commandHistoryPosition != -1) {
...
 
+	private void showFilePicker(){
+		FileChooser fileChooser = new FileChooser();
+		File currentSaveFile = new File(filePath);
+		fileChooser.setInitialDirectory(currentSaveFile.getParentFile());
+		fileChooser.setInitialFileName(currentSaveFile.getName());
+		
+		// Set extension filter
+        FileChooser.ExtensionFilter extFilter = new FileChooser.ExtensionFilter(
+                "JSON files (*.json)", "*.json");
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/BottomPaneController.java]: i think no need. If the user wants to make it a `txt` file, let him do it.

**Comment by tran-tien-dat** [comment]
A whole lots of things being done in this pr :cry: it's really hard to follow and test them all. Next time please make 1 pr per 1 feature only.

**Comment by tran-tien-dat** [comment]
Can you put a comment in the "Files changed" tab which part of the code is for which feature?

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java [hunk 4]
```diff
@@ -42,35 +43,36 @@ public TaskEventRowFactory(DataLists dataList, DataLists displayList, GridPane p
...
-	public void showAllTaskEvents(){
+
+	public int showAllTaskEvents(){
 		LocalDate currentDate = null;
 		int eventCounter = 0;
 		int deadlineTaskCounter = 0;
-		
+		int addedCounter = 0;
+
 		List<Event> eventsList = masterList.getEventsList();
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: can you rename masterList to dataList, since it's not always the masterList?

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java [hunk 5]
```diff
@@ -78,52 +80,51 @@ public void showAllTaskEvents(){
...
-		    }
-		    else if(!currentDate.equals(tempDate)){
-		    	currentDate = tempDate;
-		    	addLabel(currentDate);
-		    }
-		    
-		    addTaskEvent(itemToAdd);
+			}
+
+			LocalDate tempDate = getLatestDate(currentDate, itemToAdd);
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: I totally don't understand `getLatestDate` and `tempDate`. Shouldn't you just update `currentDate` once you've chosen `itemToAdd` above? `currentDate = nextEvent/nextDeadlineTask .getDate`

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java [hunk 6]
```diff
@@ -78,52 +80,51 @@ public void showAllTaskEvents(){
...
 		for(FloatingTask task: masterList.getFloatingTasksList()){
+			addedCounter++;
 			addTaskEvent(task);
 		}
 		
+		return addedCounter;
 	}
-	
+
 	public int showTaskEventsFromDate(LocalDate selectedDate){
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: This function is just the same as the one above without adding floating task. Shouldn't the same code be abstracted in its own function?

Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: Okay I see the diff now. You the old function you process the entire list while this one you only process half way at a certain point. Just use binary search to find the starting index that you should start working on. (java has built in binary search if i am not wrong). So have a common function like `showTaskEventsFromIndex(eventInd, deadlineTaskInd)`

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java [hunk 7]
```diff
@@ -294,7 +307,13 @@ else if(!event.getCompleted()){
 	}
 
 	private void addTaskEvent(DeadlineTask task){
-		displayList.addWithoutSorting(task);
+		int id = displayList.size('d') + 1;
+		if(displayList.contains(task)){
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: I think it's better to just check whether id == 0 instead of iterating through the list again.

Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: same for the few other occurences below

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java [hunk 8]
```diff
@@ -431,29 +457,29 @@ private boolean checkOverdue(DeadlineTask t){
 
 		return overdue;
 	}
-	
+
 	private boolean checkNotBefore(LocalDate date, TaskEvent taskEvent){
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: very ugly to keep checking null and type like this. For null, just make sure to check before calling this function. For type, i think should have 2 overloaded versions

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java (hunk 5 above)
Comment (SorataYuu) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: `getLatestDate` chooses the latest of either the currentDate or the nextEvent/nextDeadline Task date. I shifted the getting of the nextEvent/nextDeadlineTask date out because you had to check for alot of conditions so i made it another method.

The `tempDate` is to check whether the new date is same as the old one, if it is not then I'll create a new dateLabel

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java (hunk 6 above)
Comment (SorataYuu) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: You would still have to iterate through the list and compare events to find out which event starts on the date you want and then use that index :/

File: src/main/java/cs2103/v15_1j/jimjim/ui/BottomPaneController.java [hunk 9]
```diff
@@ -97,42 +105,32 @@ private void setUpCommandHistory(){
...
-		addHelpMessage("unmark {id}", 0);
-		addHelpMessage("rename {id} to {name}", 0);
-		addHelpMessage("reschedule {id} to {date} {time}", 1);
-		addHelpMessage("extend {id} to {date}", 1);
-		addHelpMessage("search {filter}, {filter} ...", 1);
-		addHelpMessage("hide search", 1);
-		addHelpMessage("help", 1);
-		addHelpMessage("undo", 1);
-		addHelpMessage("redo", 1);
+		helpPopOverController = new PopOverController("Help", 4);
```
Comment (SorataYuu) [inline src/main/java/cs2103/v15_1j/jimjim/ui/BottomPaneController.java]: Created a PopOverController to handle PopOvers instead of creating directly so that code can be reusable

File: src/main/java/cs2103/v15_1j/jimjim/ui/DayPickerPaneController.java [hunk 10]
```diff
@@ -85,10 +86,23 @@ private void setUpDayDetailScrollPane(){
...
 		dayDetailScrollPane.setHbarPolicy(ScrollPane.ScrollBarPolicy.NEVER);
 		dayDetailScrollPane.setFitToWidth(true); 
+		dayDetailScrollPane.setVbarPolicy(ScrollBarPolicy.ALWAYS);
+		dayDetailScrollPane.setOnScrollFinished(event -> checkScrollPosition());
 
 		BorderPane.setAlignment(dayDetailScrollPane, Pos.CENTER);
 		dayPickerPane.setCenter(dayDetailScrollPane);
 	}
+	
+	private void checkScrollPosition(){
```
Comment (SorataYuu) [inline src/main/java/cs2103/v15_1j/jimjim/ui/DayPickerPaneController.java]: Checks if the Scroll is at the top, if so then move a day backwards

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java (hunk 5 above)
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: Why do you need to get the later between currentDate and nextDate? To avoid repeating all the null checking stuff, you can just do initialize nextDate in the if else conditions above. For currentDate just initialize it to `LocalDate.MIN` so that you don't have to check whether it's null.

After you've had `currentDate` and `nextDate` just check whether they are the same to add the day header.

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java (hunk 6 above)
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: You will not repeat the same code in 2 different functions like currently.

I don't get what you mean by iterating through the list and compare events to find out which event starts on the date you want though. The list is sorted, so once i found the 1st event that start on the date I want (by binary search) I am sure that all the rest will start on or after the date i want and display them all, isn't it?

File: src/main/java/cs2103/v15_1j/jimjim/controller/JJController.java [hunk 11]
```diff
@@ -45,6 +45,7 @@ public String getFilePath() {
 	
 	@Override
 	public void setFilePath(String filePath) {
+		this.states.config.savePath = filePath;
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/controller/JJController.java]: Have to save this config too. And if they fail undo everything and return an error message. Maybe it's better to just create a SaveLocationCommand and execute it here.

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java [hunk 12]
```diff
@@ -81,73 +81,24 @@ public int showAllTaskEvents(){
...
 			}
 
-			if(checkNotBefore(selectedDate, itemToAdd)){
-				LocalDate tempDate = getLatestDate(currentDate, itemToAdd);
-
-				if(currentDate == null){
-					currentDate = tempDate;
-
-					if(!selectedDate.equals(currentDate)){
+			if(selectedDate.isBefore(itemToAddDate)){
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: Not just before but also on the same day right?

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java [hunk 13]
```diff
@@ -81,73 +81,24 @@ public int showAllTaskEvents(){
...
+				if(!currentDate.equals(itemToAddDate)){
+					if(currentDate.equals(LocalDate.MIN) && !selectedDate.equals(LocalDate.MIN)
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: checking equals to LocalDate.MIN is a bit hacky-ish as it does not tell the meaning. Basically you want to check whether you want to force display the selected date even if there are no items there. Just use a boolean as a function parameters. Also maintain a boolean to keep track of whether you have taken care of selectedDate

File: src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java [hunk 14]
```diff
@@ -54,6 +54,11 @@ public int showAllDeadlineTaskAndEvents(LocalDate selectedDate){
 		int eventCounter = 0;
 		int deadlineTaskCounter = 0;
 		int addedCounter = 0;
+		boolean displayDate = true;
+		
+		if(selectedDate.equals(LocalDate.MIN)){
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/ui/TaskEventRowFactory.java]: this is still checking whether the parameter is equal LocalDate.MIN. Actually a better approach will be giving passing selectedDate as null. Then here you check if selectedDate is null, initialize it to LocalDate.MIN. I just don't like using equal LocalDate.MIN

File: src/main/java/cs2103/v15_1j/jimjim/controller/JJController.java [hunk 15]
```diff
@@ -37,6 +37,17 @@ public DataLists getMasterList() {
...
 		return states.displayList;
 	}
+	
+	@Override
+	public String getFilePath() {
+		return this.states.config.savePath;
+	}
+	
+	@Override
+	public void setFilePath(String filePath) {
```
Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/controller/JJController.java]: wait you still haven't fixed this part?

Comment (tran-tien-dat) [inline src/main/java/cs2103/v15_1j/jimjim/controller/JJController.java]: should create a new SaveLocationCommand and return the feedback to UI to display if there are any errors.
